src/wrapper/objects/auto_array.rs

```diff
@@ -92,35 +90,42 @@ type_array!(jdouble, GetDoubleArrayElements, ReleaseDoubleArrayElements);
 ///
 /// AutoArray provides automatic array release through a call to appropriate
 /// Release<Type>ArrayElements when it goes out of scope.
-pub struct AutoArray<'local, T: TypeArray> {
-    obj: JObject<'local>,
+pub struct AutoArray<'a, 'local: 'a, 'other_local: 'a, T: TypeArray + 'other_local> {
```

Comment:
What's the lifetime `'a` for?

Reply:
Maybe naming this `'ptr` would be better, with docs something like this?:

```rust
/// # Lifetimes
///
/// While accessing the elements of an array there are potentially two local
/// reference frames involved and the temporary pointer to access the elements
/// is constrained to being valid within the intersection of these local frames.
///
/// - `'other_local` represents the local reference frame that `array` is owned by
/// - `'local` represents the current local reference frame (which may be stacked
///    on top of `'other_local`)
/// - `'ptr` represents the lifetime of the acquired pointer to the `array` elements.
```

Reply:
It's also notable that `AutoPrimitiveArray` and `AutoArray` should be consistent here which isn't currently the case

Reply:
I'm afraid I don't understand. The lifetime of the pointer isn't separate from the lifetime of the `AutoArray` wrapping it. The pointer is created by `AutoArray::new` and destroyed by the `AutoArray` destructor.

What's more, the lifetime `'a` comes from a lifetime parameter on `JNIEnv::get_array_elements`, which has no constraints other than that `'local` outlives it, so this lifetime can in fact be *longer* than that of the `AutoArray` itself.

Reply:
hmm I've possibly made a bit of a muddle of this `'a`/`'ptr` lifetime.

At least in the case of `AutoPrimitiveArray` it seems worth noting that previously the pointer was actually acquired _before_ calling `AutoPrimitiveArray::new()` so in some sense the lifetime was longer (though I also didn't initially add a named pointer lifetime there)

As you say the lifetime of the pointer isn't separate from the lifetime of the `AutoArray` wrapping it but I think by naming the pointer lifetime I had thought I was constraining that the `AutoArray` itself wouldn't be able to move outside the intersection between `'local` and `'other_local`. I'm a bit confused now about why I thought that was necessary though tbh.

I'll aim to drop the `'ptr` lifetime and also borrow the `JPrimitiveArray` as suggested

Reply:
okey the latest iteration removes the `'ptr` lifetime and adds an `'array` lifetime for borrowing the `JPrimitiveArray`.